pkg/controller/certificaterequests/sync.go

```diff
@@ -0,0 +1,184 @@
+	}
+	if err != nil {
+		return err
+	}
+
+	dbg.Info("ensuring issuer type matches this controller")
+
+	log = logf.WithRelatedResource(log, issuerObj.GetObjectMeta())
+
+	issuerType, err := apiutil.NameForIssuer(issuerObj)
+	if err != nil {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerNotFound, err.Error())
+		log.Error(err, "failed to obtain referenced issuer type")
+		return nil
+	}
+
+	// This CertificateRequest is not meant for us, ignore
+	if issuerType != c.issuerType {
+		c.log.WithValues(
+			logf.RelatedResourceKindKey, issuerType,
+		).V(5).Info("issuer reference type does not match controller resource kind, ignoring")
+		return nil
+	}
+
+	dbg.Info("validating CertificateRequest resource object")
+
+	el := validation.ValidateCertificateRequest(crCopy)
+	if len(el) > 0 {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, "BadConfig", "Resource validation failed: %v", el.ToAggregate())
+		return nil
+	}
+
+	if len(crCopy.Status.Certificate) > 0 {
+		dbg.Info("certificate exists so exiting sync")
+		c.setCertificateRequestStatus(crCopy)
+		return nil
+	}
+
+	i, err := c.issuerFactory.IssuerFor(issuerObj)
+	if err != nil {
+		c.recorder.Eventf(crCopy, corev1.EventTypeWarning, errorIssuerInit, "Internal error initialising issuer: %v", err)
+		return nil
+	}
+
+	// TODO: Metrics??
+
+	dbg.Info("invoking sign function as existing certificate does not exist")
+	return c.sign(ctx, crCopy, i)
+}
+
+// return an error on failure. If retrieval is succesful, the certificate data
+// will be stored in the certificate request status
+func (c *Controller) sign(ctx context.Context, cr *v1alpha1.CertificateRequest, issuer issuer.Interface) error {
+	log := logf.FromContext(ctx)
+
+	defer c.setCertificateRequestStatus(cr)
+
+	resp, err := issuer.Sign(ctx, cr)
+	if err != nil {
+		log.Error(err, "error issuing certificate request")
+		return err
+	}
+
+	// if the issuer has not returned any data, exit early
+	if resp == nil {
+		return nil
+	}
+
+	if len(resp.Certificate) > 0 {
+		cr.Status.Certificate = resp.Certificate
+		cr.Status.CA = resp.CA
+
+		c.recorder.Event(cr, corev1.EventTypeNormal, successCertificateIssued, "Certificate issued successfully")
+	}
+
+	return nil
+}
+
+// setCertificateRequestStatus will update the status subresource of the
+// certificate request.
+func (c *Controller) setCertificateRequestStatus(cr *v1alpha1.CertificateRequest) {
+	// No cert exists yet
+	if len(cr.Status.Certificate) == 0 {
+		apiutil.SetCertificateRequestCondition(cr, v1alpha1.CertificateRequestConditionReady,
+			v1alpha1.ConditionFalse, errorCertificatePending, "Certificate issuance pending")
+		return
+	}
+
+	// invalid cert
+	_, err := pki.DecodeX509CertificateBytes(cr.Status.Certificate)
+	if err != nil {
+		apiutil.SetCertificateRequestCondition(cr, v1alpha1.CertificateRequestConditionReady,
+			v1alpha1.ConditionFalse, errorCertificateFailed, "Failed to decode certificate PEM")
+		return
```

Comment:
In order to prevent infinite re-issuance, we should set the invalid data here at the very least. Otherwise, we should check for `errorCertificateFailed` as a condition state and exit early. I'm not too sure which option is preferred...

Reply:
As in check status early in the sync?